provision/kubernetes/logs.go

```diff
@@ -0,0 +1,318 @@
+// Copyright 2020 tsuru authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package kubernetes
+
+import (
+	"bufio"
+	"context"
+	"fmt"
+	"sort"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/tsuru/tsuru/provision"
+	appTypes "github.com/tsuru/tsuru/types/app"
+	apiv1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	knet "k8s.io/apimachinery/pkg/util/net"
+)
+
+const (
+	logLineTimeSeparator = " "
+	logWatchBufferSize   = 1000
+)
+
+var (
+	watchNewPodsInterval = time.Second * 10
+	watchTimeout         = time.Hour
+)
+
+func (p *kubernetesProvisioner) ListLogs(app appTypes.App, args appTypes.ListLogArgs) ([]appTypes.Applog, error) {
+	clusterClient, err := clusterForPool(app.GetPool())
+	if err != nil {
+		return nil, err
+	}
+	if !clusterClient.LogsFromAPIServerEnabled() {
+		return nil, provision.ErrLogsUnavailable
+	}
+	clusterController, err := getClusterController(p, clusterClient)
+	if err != nil {
+		return nil, err
+	}
+
+	ns, err := clusterClient.AppNamespace(app)
+	if err != nil {
+		return nil, err
+	}
+
+	podInformer, err := clusterController.getPodInformer()
+	if err != nil {
+		return nil, err
+	}
+
+	pods, err := podInformer.Lister().Pods(ns).List(listPodsSelectorForLog(args))
+	return listLogsFromPods(clusterClient, ns, pods, args)
+}
+
+func (p *kubernetesProvisioner) WatchLogs(app appTypes.App, args appTypes.ListLogArgs) (appTypes.LogWatcher, error) {
+	clusterClient, err := clusterForPool(app.GetPool())
+	if err != nil {
+		return nil, err
+	}
+	if !clusterClient.LogsFromAPIServerEnabled() {
+		return nil, provision.ErrLogsUnavailable
+	}
+	clusterClient.SetTimeout(watchTimeout)
+
+	clusterController, err := getClusterController(p, clusterClient)
+	if err != nil {
+		return nil, err
+	}
+
+	ns, err := clusterClient.AppNamespace(app)
+	if err != nil {
+		return nil, err
+	}
+
+	podInformer, err := clusterController.getPodInformer()
+	if err != nil {
+		return nil, err
+	}
+
+	selector := listPodsSelectorForLog(args)
+	pods, err := podInformer.Lister().Pods(ns).List(selector)
+	if err != nil {
+		return nil, err
+	}
+	watchingPods := map[string]bool{}
+	ctx, done := context.WithCancel(context.Background())
+	watcher := &k8sLogsNotifier{
+		ctx:           ctx,
+		ch:            make(chan appTypes.Applog, logWatchBufferSize),
+		ns:            ns,
+		clusterClient: clusterClient,
+		done:          done,
+	}
+	for _, pod := range pods {
+		if pod.Status.Phase == apiv1.PodPending {
+			continue
+		}
+		watchingPods[pod.ObjectMeta.Name] = true
+		go watcher.watchPod(pod, false)
+	}
+
+	podListener := &podListener{
+		OnEvent: func(pod *apiv1.Pod) {
+			_, alreadyWatching := watchingPods[pod.ObjectMeta.Name]
+
+			if !alreadyWatching && pod.Status.Phase != apiv1.PodPending {
+				watchingPods[pod.ObjectMeta.Name] = true
+				go watcher.watchPod(pod, true)
+			}
+		},
+	}
+
+	clusterController.addPodListener(args.AppName, podListener)
```

Comment:
Instead of calling `addPodListener` with a `*podListener` what do you think about having it declared as:
```
func (controller) addPodListener(appName string, listener interface{ OnEvent(*apiv1.Pod) })
```
And then implementing the `OnEvent` handler in the `k8sLogsNotifier` itself. This way we would avoid this closure here and simply call:
```
clusterController.addPodListener(args.appName, watcher)
```